hasktorch/src/Torch/Static/Native.hs

```diff
@@ -836,21 +955,58 @@ is_signed _self = unsafePerformIO $ (cast1 ATen.is_signed_t) _self
 --fbgemm_is_cpu_supported  = unsafePerformIO $ (cast0 ATen.fbgemm_is_cpu_supported) 
 
 -- |
--- >>> dtype &&& shape $ log (ones :: Tensor Float '[3,2])
+-- >>> dtype &&& shape $ log (ones :: Tensor 'D.Float '[3,2])
 -- (Float,[3,2])
 log :: Tensor dtype shape -> Tensor dtype shape
 log _self = unsafePerformIO $ (cast1 ATen.log_t) _self
 
 -- |
--- >>> dtype &&& shape $ logdet (ones :: Tensor Float '[2,2])
+-- >>> dtype &&& shape $ logdet (ones :: Tensor 'D.Float '[2,2])
 -- (Float,[])
--- >>> dtype &&& shape $ logdet (ones :: Tensor Float '[3,2,2])
+-- >>> dtype &&& shape $ logdet (ones :: Tensor 'D.Float '[3,2,2])
 -- (Float,[3])
 logdet :: Tensor dtype shape -> Tensor dtype (Det shape)
 logdet _self = unsafePerformIO $ (cast1 ATen.logdet_t) _self
 
--- logsumexp :: Tensor dtype shape -> Int -> Bool -> Tensor dtype shape
--- logsumexp _self _dim _keepdim = unsafePerformIO $ (cast3 ATen.logsumexp_tlb) _self _dim _keepdim
+type family IsFloatingPoint (dtype :: D.DType) :: Constraint where
+  IsFloatingPoint 'D.Half   = ()
+  IsFloatingPoint 'D.Float  = ()
+  IsFloatingPoint 'D.Double = ()
+  IsFloatingPoint dtype  = TypeError (Text "Data type " :<>:
+                                      ShowType dtype :<>:
+                                      Text " not supported")
+
+type family IsIntegral (dtype :: D.DType) :: Constraint where
+  IsIntegral 'D.Bool = ()
+  IsIntegral 'D.UInt8 = ()
+  IsIntegral 'D.Int8 = ()
+  IsIntegral 'D.Int16 = ()
+  IsIntegral 'D.Int32 = ()
+  IsIntegral 'D.Int64 = ()
+  IsIntegral dtype  = TypeError (Text "Data type " :<>:
+                                 ShowType dtype :<>:
+                                 Text " not supported")                         
+
+-- | See https://pytorch.org/docs/stable/torch.html#torch.logsumexp.
+-- >>> t = fromJust [[5, 1], [3, 2], [4, 1], [2, 7]] :: Tensor 'D.Float '[4, 2]
+--
+-- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [Float]) $ (logsumexp @1 @DropDim t :: Tensor 'D.Float '[4])
+-- (Float,([4],[5.01815,3.3132617,4.0485873,7.0067153]))
+--
+-- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[Float]]) $ (logsumexp @1 @KeepDim t :: Tensor 'D.Float '[4, 1])
+-- (Float,([4,1],[[5.01815],[3.3132617],[4.0485873],[7.0067153]]))
+--
+-- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [Float]) $ (logsumexp @0 @DropDim t :: Tensor 'D.Float '[2])
+-- (Float,([2],[5.44019,7.0116277]))
+--
+-- >>> dtype &&& shape &&& (\t' -> D.asValue (toDynamic t') :: [[Float]]) $ (logsumexp @0 @KeepDim t :: Tensor 'D.Float '[1, 2])
+-- (Float,([1,2],[[5.44019,7.0116277]]))
+logsumexp
```

Comment:
logsumexp throws for integral tensors, hence the type-level constraint to floating point tensors